ys001-yggdrasil-core-specification.md

```diff
@@ -0,0 +1,861 @@
+predecessor.
+
+### Requests
+
+At any point, a node may send a DHT request for a given Node ID. The DHT request
+**must** contain the following:
+
+1. The current coordinates of the requesting node
+1. The known bits of the target Node ID (where all unknown bits **must** be set
+   to `0`)
+
+The DHT request is then sent to the locally-known/cached DHT neighbour that has
+the closest Node ID to the target Node ID in the request. Note that a request is
+not guaranteed to yield a response, therefore the node **should** time out
+waiting for a response after a given timeframe.
+
+### Responses
+
+Having received a DHT request, a node **must** return a DHT response
+containing:
+
+1. The coordinates and public encryption key of the responding node
+1. The target Node ID, exactly as it was sent in the original DHT request
+1. Zero or more sets of candidate public encryption keys and coordinates,
+   referring to both the closest and the furthest away nodes that the node knows
+   about to the target Node ID
+
+In this context, "closest" is defined as the successor of the target Node ID
+(the node which appears at or immediately after the target Node ID in the ring),
+and "furthest" is defined to be the predecessor (the node which appears immediately
+before the target in the ring). This definition of "closest" is required to ensure
+that a partial Node ID with unknown bits set to 0 is owned by a node with the full
+Node ID, which would not be the case if the keyspace distance metric was reversed.
+Nodes **may** search with the keyspace distance metric reversed, but **must**
+set unknown bits of the Node ID to `1` in such a case.
+
+### Searches
+
+To look up a specific node on the network, a search is started by sending a DHT
+request targeting a specific Node ID.
+
+The requesting node **must** keep a record of each Node ID that is being
+searched for, until a time that either the search has completed, either by
+contacting a node with a key which matches the known bits of the target Node ID
+or by running out of nodes to check.
+
+The node will then:
+
+1. Receive a DHT response containing information about zero or more closer nodes
+   to the target Node ID
+1. Receive no response at all, at which point the search should time out
+
+Once the requesting node receives the DHT response from the responding node, the
+requesting node **must** verify that the Node ID in the DHT response matches a
+search that we initiated. If it does not match an existing search, the response
+**must** be dropped and ignored.
+
+If the search is valid, the node **must** then check if the Node ID generated
+from the sender's public encryption key in the protocol message headers matches
+the searched target Node ID (after applying a bitmask to account for any
+unknown bits of the Node ID). If it does, the search is considered to be
+completed and the node **should not** send any additional search requests
```

Comment:
They're cached indefinitely? That doesn't sound too good, there should be limit of how long data is cached.

Reply:
Review is outdated - which lines are you referring to exactly?